### ISSUE TITLE
Can't create two lambdas using the same eventBus

### ISSUE THREAD

**Issue by eduardonunesp** [serverless/serverless#6882]
# Bug Report

## Description

Create two lambda functions using the same eventBus on EventBridge service gives an error that eventBus cannot be created.

1. What did you do?
Try to create two lambda functions using the same event bus

1. What happened?
Gives an error failed to create resource already exists

1. What should've happened?
Create the two lambdas using the same eventBus

1. What's the content of your `serverless.yml` file?
```yaml
service: my-service

provider:
  name: aws
  stage: ${opt:stage}
  runtime: go1.x
  region: us-east-1

package:
  exclude:
  stage: ${opt:stage}
    - ./**
  include:
    - ./bin/**

functions:
  plans-create:
    handler: bin/eventbridge/plans/create
    events:
      - eventBridge:
          eventBus: mybus-${opt:stage}
          pattern:
            source:
            - prod.admin.plans
            detail-type:
            - plan-created

  plans-update:
    handler: bin/eventbridge/plans/update
    events:
      - eventBridge:
          eventBus: mybus-${opt:stage}
          pattern:
            source:
              - prod.admin.plans
            detail-type:
              - plan-updated
``` 

 Failed to create resource. Event bus mybus-dev already exists.

Similar or dependent issues:

- #1842

**Comment by eduardonunesp** [comment]
Ok find a way to workaround this, like the following example, but it needs at least the `arn:aws` to work

```yaml
custom:
  accountNumber: 123456789
  eventBus: arn:aws:events:us-east-1:${self:custom.accountNumber}:event-bus/mybus-${opt:stage}

functions:
  plans-create:
    handler: bin/eventbridge/plans/create
    events:
      - eventBridge:
          eventBus: ${self:custom.eventBus}
          pattern:
            source:
            - prod.admin.plans
            detail-type:
            - plan-created

resources:
  Resources:
    EventBridgeBus:
      Type: AWS::Events::EventBus
      Properties:
          Name: mybus-${opt:stage}
```

**Comment by eduardonunesp** [comment]
The downside is that you can't delete the bus later since it should have some rules on it

**Comment by TreeMan360** [comment]
I have the same issue FWIW

**Comment by TreeMan360** [comment]
Is anyone looking at this? It’s causing me an issue in combination with https://github.com/horike37/serverless-step-functions/issues/289#issuecomment-568885886

**Comment by michalorman** [comment]
Having same issue. I was trying to work this around with following:

```yaml
functions:
  function-name:
    handler: ...
    events:
      - eventBridge:
          eventBus: !GetAtt
            - EventBus
            - Arn

resources:
  Resources:
    EventBus:
      Type: AWS::Events::EventBus
      Properties:
        Name: custom-name
```

And getting:

```
TypeError: EventBus.startsWith is not a function
      at /Users/morman/.nvm/versions/node/v12.14.0/lib/node_modules/serverless/lib/plugins/aws/package/compile/events/eventBridge/index.js:123:30
```

I'd expect that the `EventBus` created in `resources` section can be properly referenced and it's arn obtained with `!GetAtt`.

I'm on 1.62 version:

```
$ sls -v
Framework Core: 1.62.0
Plugin: 3.3.0
SDK: 2.3.0
Components Core: 1.1.2
Components CLI: 1.4.0
```

**Comment by TreeMan360** [comment]
@michalorman the creation of the bus resource seems fine to me. But you need to get the ARN with: `Fn::GetAtt: [EventBus, Arn]`

I hope that sorts your issue out for you?

**Comment by michalorman** [comment]
@TreeMan360 Got the same  `EventBus.startsWith is not a function` error with:

```json
    events:
      - eventBridge:
          eventBus:
            Fn::GetAtt: [EventBus, Arn]
```

The event bus is being created fine when using resources, but I can't reference with `GetAtt`.

So far I'm having 2 issues:

1. Creating 2 lambdas using same eventBus with a custom name (according to serverless documentation it automatically creates event bus in such case) raises Cloud Formation error that EventBus is already created.
2. Creating custom eventBus in resources section I can't reference it with `!GetAtt` nor `Fn::GetAtt`.

I can work this around with the custom attribute hardcoding the ARN as Eduardo did in this thread, but would be good to do it in more declarative way rather than hardcoding ARN's.

**Comment by TreeMan360** [comment]
@michalorman this is my serverless.yml for my infrastructure project. It has a lot of other stuff in there but I am on mobile and don’t have time to split it up (cut and paste PITA 🤪)

``` plugins:
  - serverless-pseudo-parameters
  - serverless-localstack
  - serverless-offline
  - serverless-deployment-bucket

provider:
  name: aws
  runtime: nodejs12.x
  region: eu-west-1
  stage: local
  deploymentBucket:
    name: ${self:service}-${self:custom.currentStage}-deployment-bucket
    serverSideEncryption: AES256

custom:
  identifier: ${self:service}:${self:custom.currentStage}
  currentStage: ${opt:stage, self:provider.stage}
  localstack:
    stages:
      - local
    autostart: false
    debug: true
    host: http://localhost
    lambda:
      mountCode: true
    docker:
      sudo: false
  RDS:
    DB_NAME: db8dc${self:custom.currentStage}
    DB_CLUSTER_IDENTIFIER: db-cluster-8dc-${self:custom.currentStage}
    USERNAME: master
    HOST:
      Fn::GetAtt: [DBCluster, Endpoint.Address]
    ARN: arn:aws:rds:${self:provider.region}:#{AWS::AccountId}:cluster:${self:custom.RDS.DB_CLUSTER_IDENTIFIER}
    SECRET_NAME: secrets-${self:custom.RDS.DB_NAME}
    SECRET_ATTACHMENT_ARN:
      Ref: DBSecretClusterAttachment
    VPC_CIDR: 10
    ScalingConfiguration:
      prod:
        AutoPause: false
        MinCapacity: 1
        MaxCapacity: 64
      default:
        AutoPause: false
        MinCapacity: 1
        MaxCapacity: 1
        # SecondsUntilAutoPause: 900 # 15 min
  S3:
    UserFiles:
      ARN:
        Fn::GetAtt: [S3BucketUserFiles, Arn]
      BUCKET_NAME:
        Ref: S3BucketUserFiles
    BackendEvents:
      ARN:
        Fn::GetAtt: [S3BucketBackendEvents, Arn]
      BUCKET_NAME:
        Ref: S3BucketBackendEvents
      ROLE_ARN:
        Fn::GetAtt: [S3BucketBackendEventsRole, Arn]
      STREAM_ARN:
        Fn::GetAtt: [FirehoseDataLakeBackendEventsStream, Arn]
  EventBridge:
    EventBus:
      NAME:
        Ref: EventBus
      ARN:
        Fn::GetAtt: [EventBus, Arn]
resources:
  Resources:
    S3BucketUserFiles:
      Type: AWS::S3::Bucket
      Properties:
        AccessControl: PublicRead
        BucketName: 8dc-user-files-${self:custom.currentStage}
    S3BucketBackendEvents:
      Type: AWS::S3::Bucket
      Properties:
        AccessControl: Private
        BucketName: 8dc-backend-events-${self:custom.currentStage}
    S3BucketBackendEventsRole:
      Type: AWS::IAM::Role
      Properties:
        AssumeRolePolicyDocument:
          Version: '2012-10-17'
          Statement:
            - Effect: Allow
              Action:
                - sts:AssumeRole
              Principal:
                Service: firehose.amazonaws.com
        Policies:
          - PolicyName: root
            PolicyDocument:
              Version: '2012-10-17'
              Statement:
                - Effect: Allow
                  Action:
                    - 's3:*'
                  Resource: ${self:custom.S3.BackendEvents.ARN}
    DBSecret:
      Type: AWS::SecretsManager::Secret
      Properties:
        Name: ${self:custom.RDS.SECRET_NAME}
        Description: RDS database auto-generated user password
        GenerateSecretString:
          SecretStringTemplate: '{"username": "${self:custom.RDS.USERNAME}"}'
          GenerateStringKey: 'password'
          PasswordLength: 30
          ExcludeCharacters: '"@/'
    DBCluster:
      Type: AWS::RDS::DBCluster
      Properties:
        DBClusterIdentifier: ${self:custom.RDS.DB_CLUSTER_IDENTIFIER}
        DatabaseName: ${self:custom.RDS.DB_NAME}
        MasterUsername:
          Fn::Join:
            - ''
            - - '{{resolve:secretsmanager:'
              - Ref: DBSecret
              - ':SecretString:username}}'
        MasterUserPassword:
          Fn::Join:
            - ''
            - - '{{resolve:secretsmanager:'
              - Ref: DBSecret
              - ':SecretString:password}}'
        Engine: aurora
        EngineMode: serverless
        EngineVersion: 5.6.10a
        ScalingConfiguration: ${self:custom.RDS.ScalingConfiguration.${self:custom.currentStage}, self:custom.RDS.ScalingConfiguration.default}
    DBSecretClusterAttachment:
      Type: AWS::SecretsManager::SecretTargetAttachment
      Properties:
        SecretId:
          Ref: DBSecret
        TargetId:
          Ref: DBCluster
        TargetType: AWS::RDS::DBCluster
    FirehoseDataLakeBackendEventsStream:
      Type: AWS::KinesisFirehose::DeliveryStream
      Properties:
        DeliveryStreamName: 8dc-backend-events-${self:custom.currentStage}
        DeliveryStreamType: DirectPut
        S3DestinationConfiguration:
          BucketARN: ${self:custom.S3.BackendEvents.ARN}
          BufferingHints:
            IntervalInSeconds: 60
            SizeInMBs: 1
          CompressionFormat: 'UNCOMPRESSED'
          RoleARN: ${self:custom.S3.BackendEvents.ROLE_ARN}
    EventBus:
      Type: AWS::Events::EventBus
      Properties:
        Name: 8dc-${self:custom.currentStage}
    EventBusDataLakeRule:
      Type: AWS::Events::Rule
      Properties:
        Description: Sends events to Firehose for S3 data lake ingestion
        EventBusName: ${self:custom.EventBridge.EventBus.NAME}
        EventPattern:
          source:
            - accounts
            - meetings
            - payments
            - notifications
            - system
        Targets:
          - Id: match-all-events-data-lake-event-ingestion-${self:custom.currentStage}
            Arn: ${self:custom.S3.BackendEvents.STREAM_ARN}
            RoleArn: ${self:custom.S3.BackendEvents.ROLE_ARN}
  Outputs:
    DatabaseName:
      Description: RDS Database Name
      Value: ${self:custom.RDS.DB_NAME}
      Export:
        Name: ${self:custom.identifier}-DatabaseName
    DatabaseArn:
      Description: RDS Database ARN
      Value: ${self:custom.RDS.ARN}
      Export:
        Name: ${self:custom.identifier}-DatabaseArn
    SecretArn:
      Description: RDS Secret ARN
      Value: ${self:custom.RDS.SECRET_ATTACHMENT_ARN}
      Export:
        Name: ${self:custom.identifier}-SecretArn
    S3UserFilesBucketName:
      Description: S3 User Files Bucket Name
      Value: ${self:custom.S3.UserFiles.BUCKET_NAME}
      Export:
        Name: ${self:custom.identifier}-S3UserFilesBucketName
    S3UserFilesArn:
      Description: S3 User Files ARN
      Value: ${self:custom.S3.UserFiles.ARN}
      Export:
        Name: ${self:custom.identifier}-S3UserFilesArn
    S3BackendEventsBucketName:
      Description: S3 Backend Events Bucket Name
      Value: ${self:custom.S3.BackendEvents.BUCKET_NAME}
      Export:
        Name: ${self:custom.identifier}-S3BackendEventsBucketName
    S3BackendEventsArn:
      Description: S3 Backend Events ARN
      Value: ${self:custom.S3.BackendEvents.ARN}
      Export:
        Name: ${self:custom.identifier}-S3BackendEventsArn
    EventBridgeEventBusName:
      Description: 8dc Custom EventBridge EventBus Name
      Value: ${self:custom.EventBridge.EventBus.NAME}
      Export:
        Name: ${self:custom.identifier}-EventBridgeEventBusName
    EventBridgeEventBusArn:
      Description: 8dc Custom EventBridge EventBus ARN
      Value: ${self:custom.EventBridge.EventBus.ARN}
      Export:
        Name: ${self:custom.identifier}-EventBridgeEventBusArn
```

**Comment by stuartforrest-infinity** [comment]
Is this related to #7063

**Comment by DaveLo** [comment]
> Is this related to #7063?

The ` .startsWith` error is the same for Ref as for ImportValue yeah. It has to do with how the EventBridge event compiler is checking for an existing bus.  I'm not sure about the original  issue for this ticket though

> Failed to create resource. Event bus mybus-dev already exists.

That seems likely a different, but potentially related issue (hard to say).

**Comment by jackcbrown89** [comment]
@michalorman Please create a separate issue for https://github.com/serverless/serverless/issues/6882#issuecomment-581040169. I'm running into the same problem.

**Comment by kyledmellander** [comment]
This is still an ongoing issue. Using the workaround from @eduardonunesp got us unblocked.\

**Comment by uglow** [comment]
I know this framework is "free", but I would have thought having 2 Lambdas on the same Event Bus was a *very* common use-case (and test case). Building an event bus for a single subscriber would be the edge-case here. Unfortunately I don't have the knowledge to create a PR to fix this issue. But I sure would appreciate it - as would most people that use EventBridge - if this issue could be addressed. Thanks 🙏 .

**Comment by XM37** [comment]
Is a bugfix already planned here?

**Comment by stuartforrest-infinity** [comment]
@XM37 I have some work in progress that should address this attached to #7063. I have have had to pause on finishing/testing it at the moment though.

**Comment by tomfloresa** [comment]
Seems to be this is still an issue. @stuartforrest-infinity did you manage to do something?
Seriously thinking about creating the EventBridge resource through terraform and just pass ARN to serverless.yml in order to keep this piece of infrastructure centralized in tf.

**Comment by dguisinger** [comment]
Its 2021.  Any progress?  Its hard to even give it the excuse that this framework is free like @uglow said, as Serverless has paid plan.  It looks like we are coming up on 1 1/2 years of issues surrounding EventBridge as a trigger source.

**Comment by TreeMan360** [comment]
I use EventBridge extensively using Serverless and have been for some time and I don’t have any issues as long as I put the EventBridge definition in another stack.

I create an infrastructure stack (infrastructure.yml) and then I reference the Event Bus Name and Arn as outputs of that stack. Then the bus already exists and I have no issues.

In reality most of my projects have several serverless stacks so having an infrastructure stack normally becomes necessary for me at some point anyway. And I prefer it like this. 

There is definitely an issue using EB in a single stack. I tried it again a couple of weeks ago and it does not work properly. So yes it should be addressed, but for now just split your stacks and everything will be awesome 🤩.

I did elude to this ages ago above. I hope this helps someone.

**Comment by crsepulv** [comment]
I created the event bus on the console,not in serverles.yml.
Trying to deploy prod and dev  I got an error that the event is already created, but I'm not creating it just referencing.

@eduardonunesp 's soluions works for me. 
Define a custom variable with bus ARN and reference that on the function.

https://github.com/serverless/serverless/issues/6882#issuecomment-546333897
